### PR TITLE
Minor: Move TableProviderFactories up out of `RuntimeEnv` and into `SessionState`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Part of https://github.com/apache/arrow-datafusion/issues/1754

# Rationale for this change
I am trying to extract the physical_plan code into its own crate; and to do so I need to remove the circular dependencies between core --> datasource --> execution --> datasource

Table factories are used for planning, not for execution, but RuntimeEnv is used for execution (and I am trying to move it into the `datafusion_execution` crate)

See more details in https://github.com/apache/arrow-datafusion/issues/1754#issuecomment-1452438453

# What changes are included in this PR?
1. Move table_factories off of RuntimeEnv and directly to SessionState

# Are these changes tested?
Covered by existing tests (and the test changes illustrate what happened to the API). I believe this will affect Ballista (@avantgardnerio ) and delta-rs (cc @roeap )

# Are there any user-facing changes?

Users who are specifying custom TableFactoryProviders have a slightly different API to register them.

### CODE REVIEW

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -278,6 +282,15 @@ impl SessionContext {
         self.session_id.clone()
     }
 
+    /// Return the [`TableFactoryProvider`] that is registered for the
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:285]:
```suggestion
    /// Return the [`TableProviderFactory`] that is registered for the
```

File: datafusion/core/src/execution/runtime_env.rs [hunk 2]
```diff
@@ -22,10 +22,7 @@ use crate::{
     error::Result,
     execution::disk_manager::{DiskManager, DiskManagerConfig},
 };
-use std::collections::HashMap;
 
-use crate::datasource::datasource::TableProviderFactory;
```
Comment (alamb) [inline datafusion/core/src/execution/runtime_env.rs:27]: The whole point of the PR is to remove the `datasource` dependency from the `runtime_env` module (as I am trying to move the `runtime_env` module into a new crate, `datafusion-execution`, that does not depend on `datasource`)

I plan to move object_store registry in a separate PR to keep the reviews smaller

File: datafusion/core/src/execution/runtime_env.rs [hunk 3]
```diff
@@ -129,24 +114,12 @@ pub struct RuntimeConfig {
...
-    /// Custom table factories for things like deltalake that are not part of core datafusion
-    pub table_factories: HashMap<String, Arc<dyn TableProviderFactory>>,
 }
 
 impl RuntimeConfig {
     /// New with default values
     pub fn new() -> Self {
-        let mut table_factories: HashMap<String, Arc<dyn TableProviderFactory>> =
-            HashMap::new();
-        table_factories.insert("PARQUET".into(), Arc::new(ListingTableFactory::new()));
```
Comment (alamb) [inline datafusion/core/src/execution/runtime_env.rs:141]: this construction now happens as part of creating SessionState

File: datafusion/core/tests/sql/create_drop.rs [hunk 4]
```diff
@@ -126,12 +129,14 @@ async fn create_custom_table() -> Result<()> {
...
 #[tokio::test]
 async fn create_external_table_with_ddl() -> Result<()> {
-    let mut cfg = RuntimeConfig::new();
-    cfg.table_factories
-        .insert("MOCKTABLE".to_string(), Arc::new(TestTableFactory {}));
+    let cfg = RuntimeConfig::new();
     let env = RuntimeEnv::new(cfg).unwrap();
     let ses = SessionConfig::new();
-    let ctx = SessionContext::with_config_rt(ses, Arc::new(env));
+    let mut state = SessionState::with_config_rt(ses, Arc::new(env));
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:135]: Here is the new pattern of how to add a table provider factory

**Review by avantgardnerio — APPROVED** [review]
Seems okay...

**Comment by alamb** [comment]
Thank you for the review @avantgardnerio
